### PR TITLE
is_alias_of changes for storageless tensors

### LINKED ISSUE #71377: is_alias_of support for storageless tensors
### 🚀 The feature, motivation and pitch

Motivation:
------------
Torch DDP code uses is_alias_of() for two tensors to check if they share the same storage. The current is_alias_of() API on the tensor checks if the underlying Storage class for both the tensors has a common storage_impl_. This is also not a virtual method, so not an overridable API for the tensor currently.
For backends like HPU, that use lazy storageless tensors and uses DDP for distributed training, the is_alias_of() API doesn't work. 

Pitch:
------
The is_alias_of() API by default should check if the tensor has storage, before comparing the storage_impl_. This should ensure for storageless tensors, the default API doesn't fail with no storage error.
Also, the is_alias_of() API needs to be virtual, so that a backend can override it and provide its own implementation for  is_alias_of(). The backend can provide a result by using its own mechanism to detect alias tensors that are storageless. For HPU lazy storageless tensors, the backend would determine if they tensors are sharing storage internally on the device and return a result accordingly.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sujoysaraswati**
The is_alias_of() method should check has_storage() before
returning true/false for storageless tensors. Also, it should be
possible to override the default behavior by making it virtual.

Fixes #71377

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
I think this LGTM but adding @zou3519 @albanD to take another look.  I can help land.

File: aten/src/ATen/core/TensorBase.h [hunk 1]
```diff
@@ -301,7 +301,7 @@ class TORCH_API TensorBase {
     return impl_->storage();
   }
   bool is_alias_of(const at::TensorBase& other) const{
-    return impl_->storage().is_alias_of(other.storage());
+    return impl_->is_alias_of(*other.unsafeGetTensorImpl());
```
Comment (albanD) [inline aten/src/ATen/core/TensorBase.h:304]: This function is used in perf sensitive places so not sure that adding the virtual call is ok here.
```cpp
return impl_->has_strorage() && impl_->storage().is_alias_of(other.storage());
```
should give you the same feature without the perf loss right?

Comment (sujoysaraswati) [reply aten/src/ATen/core/TensorBase.h:304]: @albanD for backends with lazy tensors like HPU, we still retain the view tensor notion internally so that they share the same storage with the original tensor in the device and updating a view tensor results in an update on the original tensor. However, the lazy tensors are storage less and impl_->has_storage() will always return false even if the two tensors are views.
We want a way to override the default is_alias_of so that we can define it based on the view related records we maintain in our bridge. I did notice that the is_alias_of gets used in the DDP code and making it virtual may have a slightly higher cost, but currently the DDP code seems to assume that the tensors are with storage always. Hence, I had to make it virtual so that we can override the default behavior for storage less tensors.
Is there any other suggestion from you about how this can be done for storage less tensors to return true for is_alias_of for views like gradient as bucket view cases in DDP?

Comment (albanD) [reply aten/src/ATen/core/TensorBase.h:304]: The DDP usage is not the one that worries me. It is more the use within the jit for IValues as well as the use in TensorIterator (which is the base of all our low level kernels).

If your Tensors have a concept of "storage" and actually do share them. Why not populate it here so that this kind of logic will work as expected?

Comment (sujoysaraswati) [reply aten/src/ATen/core/TensorBase.h:304]: I had noticed the usage in TorchScript IValue, but didn't notice the use in TensorIterator. Is the TorchScript JIT in the time sensitive path? There are a lot of asserts in the autograd generated ops where the storage is_alias_of() is checked if the tensor has_storage() is true. They sounded fine as the code is explicitly checking for storage presence of tensors. 
  Can you point me to the TensirIterator usage? If it is in performance sensitive code, but does not handle storageless lazy tensors, maybe there should be another API that can be used for possibly storageless tensors explicitly? Let me know your opinion. How is it going to be handled for lazy tensor core, will is_alias_of() work for them?
  For the tensors in our backend, they are lazy tensors and do not have storage. Hence, at the PyTorch at::tensor level the has_storage() is false. I was mentioning that internally, we do bookeeping for views created on tensors and on the device they share storage, but this is handled within the backend during tensor allocations and not associated with the lazy tensor.

Comment (sujoysaraswati) [reply aten/src/ATen/core/TensorBase.h:304]: We had discussions with @wconstab about the HPU lazy tensor and the current LazyTensorCore. There are few differences, including the functionalization pass, so it isn't usable for HPU in the current form.

Comment (albanD) [reply aten/src/ATen/core/TensorBase.h:304]: > Can you point me to the TensirIterator usage?

There is one in TensorIteratorReduce.cpp and MemoryOverlap.cpp (which is called to check if ops are valid IIRC).
Did you discuss with @wconstab how lazy tensor handles it?

Comment (sujoysaraswati) [reply aten/src/ATen/core/TensorBase.h:304]: @albanD , thanks - I had seen the TensorIteratorReduce.cpp, but was not sure if it is important from performance point of view as it is inside an assert. For the MemoryOverlap.cpp, it is calling the is_alias_of() from the Storage() class directly, if the storage is available. This should be unchanged with this PR, as I am changing the TensorBase is_alias_of(). Shall I modify the TensorIteratorReduce.cpp to similarly use the Storage->is_alias_of(), will it be fine?

We didn't discuss about this particular feature for LazyTensorCore with @wconstab , we had a discussion on the high level differences between HPU lazy tensor and XLA lazy tensor approach. I will discuss this with him.

**Comment by albanD** [comment]
Hey!

From more internal discussion on this, I think the main point are:
- We should rename is_alias_of to has_same_storage. This is a better description of what it does and ensure its semantic is clear.
- The places where this function is used assumes that the given Tensor is a strided Tensor. The rest of the code around it assumes that the Tensor is strided to implement its feature. So allowing this function to be virtual doesn't help if it is not properly looking like a strided Tensor.  In your case, if your Tensor is properly strided, then it should expose all these things properly (and thus wouldn't need it to be virtual).

**Comment by sujoysaraswati** [comment]
Hi @albanD, in our case we have lazy tensors and they can be strided/view tensors. We do the bookkeeping at our side to create the tensors in the device and run ops with the strided tensor semantics. However, since this is a lazy tensor, there is no storage attached so we would not be able to expose a storage for them and use has_same_storage functionality. This is why we wanted to create it as a virtual so that we can refer to our bookkeeping and return true/false based on whether they are strided or not.

Is there anything wrong to keep the strided tensor bookkeeping inside the backend bridge for a lazy tensor? It feels a bit restrictive, as we do create lazy tensors with strided/view semantics, but we can't have an API like is_alias_of that can be overridden and hence we are blocked on DDP usage.

> Hey!
> 
> From more internal discussion on this, I think the main point are:
> 
> * We should rename is_alias_of to has_same_storage. This is a better description of what it does and ensure its semantic is clear.
> * The places where this function is used assumes that the given Tensor is a strided Tensor. The rest of the code around it assumes that the Tensor is strided to implement its feature. So allowing this function to be virtual doesn't help if it is not properly looking like a strided Tensor.  In your case, if your Tensor is properly strided, then it should expose all these things properly (and thus wouldn't need it to be virtual).

**Comment by albanD** [comment]
For the DDP use in particular my understanding is the following: `is_alias_of` is only one of the functions that is being used there to make their bucketting system work. They also need advanced viewing/striding functions etc.
So you do need to provide the full "strided Tensor API" to be able to use that feature.
The fact that there is a storage, size, stride, etc are just pieces of that API. If you do have all of these, I don't see why you cannot have a Storage object here since you do have one.

Note that when I talk about Storage object here, I mean the "Storage" objects stored by your TensorImpl. This doesn't have to be the PyTorch vanilla Storage though. Advanced features like Functionalization have a special Storage class that they set on their custom TensorImpl.
In your case, you can have a LazyStorage object that will allow you to use all the Tensor methods as-is without issues.

**Comment by wconstab** [comment]
@sujoysaraswati I want to connect you with @alanwaketan from our side working on DDP design for lazy tensor core.  It's still in early stages, but we'll want to solve the same things so let's work together on a solution.

**Comment by sujoysaraswati** [comment]
> 

Thanks for the suggestions @albanD . Currently we have the lazy tensor as storageless (similar to torch/XLA), so the TensorImpl doesn't use a storage object and has_storage()/storage() will return false/raise error. I will sync up with @wconstab about the Lazy Tensor Core and how it plans to use any LazyStorage.

**Comment by sujoysaraswati** [comment]
Hi @albanD , @wconstab - we have followed your suggestion and used a dummy storage for our lazy tensor. With this, I am abandoning this PR.
